drivers: usb: USB device driver for Kinetis USBFSOTG Controller

jfischer-no opened the pull request:
USB device driver for Kinetis USBFSOTG Controller

I had a driver for Kinetis USBFSOTG controller left over here, so I adapted it for zephyr, it is still WIP/RFC. Tested with FRDM-K64F board. The LEDs/gpio code will be removed, it is only for tracing.

TODO:
- [x] it needs better configuration options for endpoints, dependent on usb class application ~~, currently 64 byte buffers are used by default~~ -> see 259e568.
- [x] ~mpu configuration (as workaround it is simply turned off :fist_oncoming: )~~
- [x] ~clock configuration~~
- [x] testing with KW22D512 (#845)
- [x] better use of both (even/odd) endpoint buffers

jfischer-phytec-iot commented:
@fvincenzo @MaureenHelm When the MPU is switched on (default on), the USBFSOTG controller signals the DMA error because it, Bus Master 4,  can not access the SRAM (BD table and the buffer are located there). As I see, I can insert a new region into nxp_mpu_regions.c and/or add BM4 Flags to MPU_REGION_xxxxx, is it correct so far?

MaureenHelm reviewed (commented):
Thanks @jfischer-phytec-iot. This looks pretty good. I wasn't able to get the cdc sample to work on frdm_k64f though; it got stuck waiting for DTR.

> @fvincenzo @MaureenHelm When the MPU is switched on (default on), the USBFSOTG controller signals the DMA error because it, Bus Master 4, can not access the SRAM (BD table and the buffer are located there). As I see, I can insert a new region into nxp_mpu_regions.c and/or add BM4 Flags to MPU_REGION_xxxxx, is it correct so far? see 678c39b

Let's add the flags to MPU_REGION_xxxx (as you've already done) to be consistent with ethernet.

jfischer-phytec-iot commented on drivers/usb/device/usb_kinetis.c:
```diff
@@ -0,0 +1,929 @@
+
+static inline void kinetis_usb_init_clock(void)
+{
+	/* TODO: refactor or use ext/mcux sdk */
```
Done, I wanted to make the clock source configurable, but the IRC48 did not work for me, so for the first the PLL is the only possible clock source.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+   usb 9-1: SerialNumber: 00.01
+   cdc_acm 9-1:1.0: ttyACM1: USB ACM device
+
+The app prints on serial output, used for the console:
```
You don't say how to build (and flash) the sample app.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+   Wait for DTR
+
+Open a serial port emulator, for example minicom
```
Don't you need to open minicom BEFORE you run the sample app (where does "Wait for DTR" mentioned above appear?)

jfischer-phytec-iot replied:
no, the sample waits for DTR, it appears after the minicom opens the device provided by the Zephyr.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+.. code-block:: console
+
+   minicom --device /dev/ttyACM1
```
default baud rate is ok?

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+Overview
+********
+
+Sample app for USB CDC ACM driver. The received data is echoed back
```
This sample app demonstrates use of a USB Communication Device Class (CDC) Abstract Control Model (ACM) driver provided by the Zephyr project.  Received data is echoed back to the serial port provided by this driver.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+Building and Running
+********************
+
+Plug the board into a host device, for example, a PC running Linux.
```
You don't say how to build (and flash) the sample app.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+   Characters read:
+
+The characters entered in serial port emulator will be echoed back.
+You may need to stop modemmanager, if it is trying to access the
```
You should mention this issue earlier in the steps -- before plugging the board to your host system I suspect.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+USB CDC ACM Sample Application
+####################################
+
```
Does this sample app only work on specific boards?  If so, you should list them here.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+************
+
+This project requires an USB device driver,
+which is available for the :ref:`arduino_101` and :ref:`frdm_k64f` boards.
```
Is this the list of supported boards then?

jfischer-phytec-iot replied:
yes, but, I've just seen that the default board for this sample is quark_se_c1000_devboard.

dbkinder commented on samples/subsys/usb/cdc_acm/README.rst:
```diff
@@ -0,0 +1,63 @@
+********************
+
+Plug the board into a host device, for example, a PC running Linux.
+The board will be detected as a CDC ACM device.
```
(as shown by the Linux `dmesg` command):

dbkinder reviewed (dismissed):
suggested edits and clarifications

jfischer-phytec-iot commented:
@dbkinder I moved 4a4c6a0 and d8c63e1 to #825

jfischer-phytec-iot commented:
rebased on master, tested with CDC ACM and DFU, usable but still WIP (leds and blocking while must be removed)

galak commented:
@jfischer-phytec-iot since still wip, added that to subject, also used github markup for the todo list.  (might need a bit of cleaning up)

jfischer-no committed:
arm: mpu: nxp: set bus master 4 to write and read access

Set bus master 4 to write and read access which allows
the USBFSOTG controller read/write access to the memory.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: usb: add usb device driver for Kinetis USBFSOTG Controller

Add usb device driver for Kinetis USBFSOTG Controller.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
arch: nxp_kinetis: enable USB device driver

Enable USB device driver for K64F and KW2xD512 SoC

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
samples: cdc_acm: enable sanitycheck for Kinetis boards

Enable sanitycheck for FRDM-K64F and USB-KW24D512 boards

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
@finikorg confirm, have seen it too, I will investigate

finikorg commented:
Need to also add this:
```
diff --git a/arch/arm/soc/nxp_kinetis/k6x/soc.h b/arch/arm/soc/nxp_kinetis/k6x/soc.h
index c6fa930c6..a509c0b6c 100644
--- a/arch/arm/soc/nxp_kinetis/k6x/soc.h
+++ b/arch/arm/soc/nxp_kinetis/k6x/soc.h
@@ -35,6 +35,7 @@ extern "C" {
 
 #define IRQ_SPI0 26
 #define IRQ_SPI1 27
+#define IRQ_USB_OTG 53
 #define IRQ_GPIO_PORTA 59
 #define IRQ_GPIO_PORTB 60
 #define IRQ_GPIO_PORTC 61

```

jfischer-phytec-iot commented:
I agreed to remove the IRQ in the hope that USB DTS support will not take so long, see #4376

jfischer-phytec-iot commented:
>[usb/usbfsotg] [ERR] usb_dc_ep_write: Endpoint is occupied by the controller

@finikorg it is so far ok, the error appears because previous IN packet was not picked up, so there is no IN token from the USB host as long as host has not configured usb ethernet device.

Unfortunately, I do not see any further write accesses for ep 0x83 and I do not know what the samples really do, with netusb FRDM-K64F has two ethernet interfaces....

finikorg commented:
@jfischer-phytec-iot So if you disable mcux ethernet on FRDM-K64F then you can ping echo_server? I was trying the same but something was not working, I will try again.

jfischer-phytec-iot commented:
@finikorg Yes, but then every second outgoing packet will be lost, I suspect IN buffer handling in the driver.

jfischer-phytec-iot commented:
@finikorg I reworked the driver, see fixup f6adb0a. Tested samples: cdc_acm, mass, webusb and http_server with the script below:

```
#!/bin/bash

IP=$1

for (( n=0; n<1; n++ ))
do
        for (( packetsize=0; packetsize<1600; packetsize=$packetsize+32 ))
        do
                echo "ping payload $packetsize"
                ping -i 0.002 -c 42 -s $packetsize -w 1 $IP
                sleep 1
        done
done

```
packetsize 40 bytes:
rtt min/avg/max/mdev = 0.611/1.216/1.275/0.103 ms

packetsize 1480 bytes:
rtt min/avg/max/mdev = 3.884/3.940/4.154/0.067 ms

jfischer-phytec-iot commented:
It is waiting for DT support, #4376

jfischer-no committed:
dts: bindings: add yaml file for Kinetis USBD support

Add yaml file for Kinetis USBD support.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
dts: arm: nxp: use DT to configure USBD on Kinetis SoC

Add DT and fixup files to configure USB device driver on
Kinetis SoC K64F and KW24D512.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
tests: samples: exclude usb_kw24d512 from wpanusb sample

Exclude usb_kw24d512 from wpanusb sample until the the sample
has been reworked and is more generic.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

finikorg reviewed (approved):
LGTM, tested CDC ACM and ECM functionality on FRDM-K64F

jfischer-phytec-iot commented:
@nashif ping

nashif commented:
> @nashif ping

@jfischer-phytec-iot pong.

@MaureenHelm have you looked at this? I assume it is ready to be re-reviewed/tested now.

MaureenHelm reviewed (commented):
LGTM, but I'd like to get someone from my USB team to also take a look.

I tested CDC on frdm_k64f and it works fine. Do the other samples work too?

jfischer-phytec-iot commented:
@MaureenHelm 
>I tested CDC on frdm_k64f and it works fine. Do the other samples work too?

Yes, this is my reference driver for usb development (netusb related samples need CONFIG_ETH_MCUX_0 to be disabled).

MaureenHelm commented:
@jfischer-phytec-iot I'm sorry for the delay, but I got some feedback from our USB team. Here's what they said:

1.  Some SOC configure code is in kinetis_usb_init(void) like:
   /* enable USB voltage regulator */ 
   SIM->SOPT1 |= SIM_SOPT1_USBREGEN_MASK;
   For different SoCs, the register to enable USB regulator might be different. This code only works for a part of Kinetis devices.
 
2. In the Kinetis USB controller, there is limitation that if the device is not ready to receive the setup token sent by the host, the device controller would hang. To avoid such issue, the driver needs to prime a transfer to receive the next setup token as soon as possible once the previous setup transfer is done. From the code we could see, the next setup prime operation would be called by the upper layer of the USB Kinetis driver, it might be not quick enough in some extreme cases.
 
3. The driver is using memcpy operation in the data transfer, which will cause huge throughput performance decrease.
 
4. dev_data.attached is set in function usb_dc_attach(), but it is not clear in usb_dc_detach
 
5. Can’t support multiple KHCI controllers case. Currently there is no multiple KHCI controllers cases in any existed NXP SoC, but from design viewpoint, it is necessary to take it into accounted.
 
6. Any STALL interrupt will cause the STALL status of EP0 be cleared. In case that both EP0 and EP1 should be stalled (sending out STALL handshake), if EP1’s STALL token is sent out and EP1’s STALL interrupt issued, the EP0’s STALL would be cleared before the STALL handshake for EP0 is sent out, then NO STALL handshake for EP0 would be sent out.
 
7. If one EP is in stall status, the transfers on this ep need to be forbidden
 
8. When the Clear_Feature(HALT) is received from the host via EP0, the data toggle needs to be reset. And any unfinished transfer needs to be cancelled. Please take the following case into accounted that the Host sends out a Clear_Feature(HALT) even the corresponding ep is NOT in stall status.

jfischer-no committed:
drivers: usb: kinetis: fixup endpoint config, stall and read

This patch fixes some bugs found during testing with testusb from
linux kernel.

An unresolved issue is that the stack is not fast enough
to stall (if necessary) the control endpoint during Setup Stage.
This might require a API change so that the usb device stack
can explicit allow the driver to resume token processing.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
@nashif @MaureenHelm @finikorg 
Updated, see last commit. I have tested it together with #7435 and the driver passes all tests.

```
./testusb -v 512 -D /dev/bus/usb/005/026                           
unknown speed   /dev/bus/usb/005/026    0
/dev/bus/usb/005/026 test 0,    0.000006 secs
/dev/bus/usb/005/026 test 1,    2.000006 secs
/dev/bus/usb/005/026 test 2,    1.999993 secs
/dev/bus/usb/005/026 test 3,    1.001027 secs
/dev/bus/usb/005/026 test 4,    1.000963 secs
/dev/bus/usb/005/026 test 5,   57.978878 secs
/dev/bus/usb/005/026 test 6,   34.993027 secs
/dev/bus/usb/005/026 test 7,   29.998934 secs
/dev/bus/usb/005/026 test 8,   18.001992 secs
/dev/bus/usb/005/026 test 9,    4.979957 secs
/dev/bus/usb/005/026 test 10,   11.952035 secs
/dev/bus/usb/005/026 test 11,   16.084910 secs
/dev/bus/usb/005/026 test 12,   16.086950 secs
/dev/bus/usb/005/026 test 13,   17.955939 secs
/dev/bus/usb/005/026 test 14,    2.464990 secs
/dev/bus/usb/005/026 test 17,    1.999973 secs
/dev/bus/usb/005/026 test 18,    2.000033 secs
/dev/bus/usb/005/026 test 19,    1.999984 secs
/dev/bus/usb/005/026 test 20,    2.000021 secs
/dev/bus/usb/005/026 test 21,    2.443951 secs
/dev/bus/usb/005/026 test 24,   55.060829 secs
/dev/bus/usb/005/026 test 27,   56.895873 secs
/dev/bus/usb/005/026 test 28,   34.153936 secs
```

Linux kernel output:
```
usbtest 5-1:1.0: TEST 0:  NOP
usbtest 5-1:1.0: TEST 1:  write 1024 bytes 1000 times                                   
usbtest 5-1:1.0: TEST 2:  read 1024 bytes 1000 times                                    
usbtest 5-1:1.0: TEST 3:  write/512 0..1024 bytes 1000 times                            
usbtest 5-1:1.0: TEST 4:  read/512 0..1024 bytes 1000 times                             
usbtest 5-1:1.0: TEST 5:  write 1000 sglists 32 entries of 1024 bytes                   
usbtest 5-1:1.0: TEST 6:  read 1000 sglists 32 entries of 1024 bytes                    
usbtest 5-1:1.0: TEST 7:  write/512 1000 sglists 32 entries 0..1024 bytes               
usbtest 5-1:1.0: TEST 8:  read/512 1000 sglists 32 entries 0..1024 bytes                
usbtest 5-1:1.0: TEST 9:  ch9 (subset) control tests, 1000 times                        
usbtest 5-1:1.0: TEST 10:  queue 32 control calls, 1000 times                           
usbtest 5-1:1.0: TEST 11:  unlink 1000 reads of 1024                                    
usbtest 5-1:1.0: TEST 12:  unlink 1000 writes of 1024                                   
usbtest 5-1:1.0: TEST 13:  set/clear 1000 halts                                         
usbtest 5-1:1.0: TEST 14:  1000 ep0out, 1..1024 vary 512                                
usbtest 5-1:1.0: TEST 17:  write odd addr 1024 bytes 1000 times core map                
usbtest 5-1:1.0: TEST 18:  read odd addr 1024 bytes 1000 times core map                 
usbtest 5-1:1.0: TEST 19:  write odd addr 1024 bytes 1000 times premapped               
usbtest 5-1:1.0: TEST 20:  read odd addr 1024 bytes 1000 times premapped                
usbtest 5-1:1.0: TEST 21:  1000 ep0out odd addr, 1..1024 vary 512                       
usbtest 5-1:1.0: TEST 24:  unlink from 1000 queues of 32 1024-byte writes               
usbtest 5-1:1.0: TEST 27: bulk write 31Mbytes                                           
usbtest 5-1:1.0: TEST 28: bulk read 31Mbytes     
```
